### PR TITLE
feat: Implement basic version of RLIKE

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

RLIKE is pretty common in ETL jobs.

## What changes are included in this PR?

This PR implements the RLIKE expression, but falls back to Spark for all regexp patterns that are not currently guaranteed to be compatible (which is all of them for now) unless a new config (`spark.comet.regexp.allowIncompatible`) is enabled.

## How are these changes tested?

New tests

### CODE REVIEW

File: spark/src/main/scala/org/apache/comet/expressions/RegExp.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+object RegExp {
+
+  /** Determine whether the regexp pattern is supported natively and compatible with Spark */
+  def isSupportedPattern(pattern: String): Boolean = {
+    // this is a placeholder for implementing logic to determine if the pattern
+    // is known to be compatible with Spark, so that we can enable regexp automatically
+    // for common cases and fallback to Spark for more complex cases
+    false
+  }
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/expressions/RegExp.scala:30]: I plan on implementing this in a future PR

**Comment by kazuyukitanimura** [comment]
Last time, we had a discussion around dictionary https://github.com/apache/datafusion-comet/pull/469#discussion_r1631685596
Was it addressed?

**Comment by andygrove** [comment]
> Last time, we had a discussion around dictionary [#469 (comment)](https://github.com/apache/datafusion-comet/pull/469#discussion_r1631685596) Was it addressed?

Yes, the test now covers the dictionary case (as far as I can tell). The expression is never invoked with a dictionary though , so it seems to be converted before reaching the expression.

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 2]
```diff
@@ -617,6 +618,136 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     }
   }
 
+  test("rlike simple case") {
+    val table = "rlike_names"
+    val gen = new DataGenerator(new Random(42))
+    Seq(false, true).foreach { withDictionary =>
+      val data = Seq("James Smith", "Michael Rose", "Rames Rose", "Rames rose") ++
+        gen.generateStrings(100, "rames Rose", 12)
```
Comment (kazuyukitanimura) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:626]: I think we need enough number of repeated string in order to make a dictionary. Right now, it has a low chance due to the randomness...

Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: @kazuyukitanimura I updated the test and have now added dictionary support

**Comment by andygrove** [comment]
Performance is ok, but not great.

```
OpenJDK 64-Bit Server VM 11.0.24+8-post-Ubuntu-1ubuntu322.04 on Linux 6.5.0-41-generic
AMD Ryzen 9 7950X3D 16-Core Processor
TPCDS Micro Benchmarks:                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
rlike                                               105            110           3          1.9         515.6       1.0X
rlike: Comet (Scan)                                 105            108           2          2.0         512.3       1.0X
rlike: Comet (Scan, Exec)                            98            114          21          2.1         482.6       1.1X
```

**Review by kazuyukitanimura — APPROVED** [review]
LGTM

**Pull request merged**